Add a subset of API-compatible endpoints to establish the feasibility of switching away from upstream BSS without losing functionality

alexlovelltroy opened the pull request:
This pull request establishes the handlers and storage functions to allow cloud-init to function as a bss server in simpole cases.  Below are the most important changes grouped by theme.

### Storage Backend Enhancements:
* Introduced a new `quackstore` backend for persistent storage, configurable via the `--storage-backend` flag. Default remains `memstore`. Added support for specifying a database file path for `quackstore` via the `--db-path` flag. (`cmd/cloud-init-server/main.go`, [[1]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918L18-R20) [[2]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918R54-R55) [[3]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918R75-R76) [[4]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918L118-R148)
* Updated imports to use `pkg/smdclient` instead of `internal/smdclient` for better modularity. (`cmd/cloud-init-server/handlers.go`, [[1]](diffhunk://#diff-db34fea3f71d4d91a800342607364173da530571390e49168b3c43193ca1e079L10-R11); `cmd/cloud-init-server/main.go`, [[2]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918L18-R20); other handler files)

### Boot Parameter Management:
* Added a new `pkg/bss` package to handle boot parameters (`BootParams` and `BootParamsV1` structs) with JSON parsing, validation, and merging capabilities. (`pkg/bss/bootparams.go`, [pkg/bss/bootparams.goR1-R112](diffhunk://#diff-06cd9112813ea3d17a319d0302549feddf68cc0407ed066c20b3f402ba44a5aeR1-R112))
* Implemented a utility function to generate iPXE boot scripts based on boot parameters, including support for root filesystem and cloud-init configurations. (`pkg/bss/bootscript.go`, [pkg/bss/bootscript.goR1-R88](diffhunk://#diff-153676849623f1183f0d08e85fa780354dee9761cb59bbbb5472b38a87506861R1-R88))

### Testing Improvements:
* Added unit tests for the `memstore` backend to validate its functionality. (`internal/memstore/ciMemStore_test.go`, [internal/memstore/ciMemStore_test.goR1-R20](diffhunk://#diff-06b74409fdcff38d41eda23dd38191894960cb97fa0d0ab0adcbf475929bae0cR1-R20))
* Created comprehensive test cases for boot parameter parsing, validation, and merging logic. (`pkg/bss/bootparams_test.go`, [pkg/bss/bootparams_test.goR1-R285](diffhunk://#diff-b92ed0fe092b298aecfa3105c4faf5c1e39aef6367505c161b341965ed2afc38R1-R285))

### Code Quality Enhancements:
* Improved the `generateInstanceID` function in `quackstore` to generate unique instance IDs using random hexadecimal values. (`internal/quackstore/quackstore.go`, [internal/quackstore/quackstore.goL351-R357](diffhunk://#diff-1362e48332f97a725f4ceea06ae862cb44f34391c3d85b473e80877f8983fb79L351-R357))
* Added new dependencies to `go.mod` for UUID generation and YAML parsing. (`go.mod`, [[1]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R10) [[2]](diffhunk://#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R55-R56)

These changes enhance the flexibility, reliability, and test coverage of the `cloud-init-server` project, paving the way for improved storage options and boot parameter handling.

alexlovelltroy committed:
feat(quackstore): implement QuackStore for persistent storage with tests

- Added QuackStore implementation for managing groups, instances, and cluster defaults using Quack for persistence.
- Introduced unit tests for QuackStore to ensure functionality and data integrity.
- Updated go.mod to include new dependencies and replaced the quack module path.
- Enhanced CloudConfigFile JSON marshaling and unmarshaling for better encoding handling.

alexlovelltroy committed:
feat(quackstore): implement QuackStore for persistent storage with tests

- Added QuackStore implementation for managing groups, instances, and cluster defaults using Quack for persistence.
- Introduced unit tests for QuackStore to ensure functionality and data integrity.
- Updated go.mod to include new dependencies and replaced the quack module path.
- Enhanced CloudConfigFile JSON marshaling and unmarshaling for better encoding handling.

alexlovelltroy committed:
feat(quackstore): implement QuackStore for persistent storage with tests

- Added QuackStore implementation for managing groups, instances, and cluster defaults using Quack for persistence.
- Introduced unit tests for QuackStore to ensure functionality and data integrity.
- Updated go.mod to include new dependencies and replaced the quack module path.
- Enhanced CloudConfigFile JSON marshaling and unmarshaling for better encoding handling.

alexlovelltroy committed:
test(memstore): add unit tests for MemStore implementation

- Introduced a new test file for MemStore to validate its functionality.
- Integrated standard test suite from cistore testing package to ensure comprehensive testing.

alexlovelltroy committed:
feat(cloud-init-server): add support for multiple storage backends

- Introduced a new command-line flag for selecting the storage backend (mem or quack).
- Default storage backend set to memstore, with a new option for quackstore using a specified database path.
- Updated the datastore initialization logic to handle both memstore and quackstore, including error handling for quackstore initialization.

synackd committed:
fix: correct UnmarshalJSON and add MarshalJSON for CloudConfigFile

synackd committed:
fix: correct UnmarshalJSON and add MarshalJSON for CloudConfigFile

synackd committed:
chore: clarify comments in custom marshal/unmarshal funcs

synackd committed:
chore: clarify comments in custom marshal/unmarshal funcs

alexlovelltroy committed:
test: add unit tests for CloudConfigFile JSON marshal/unmarshal

synackd committed:
fix(tests): use base64-encoded data for base64 marshal test

alexlovelltroy committed:
refactor: Use an alternate Unmarshal implementation

alexlovelltroy committed:
chore: add base64 encoding import for future functionality

alexlovelltroy committed:
refactor: simplify UnmarshalJSON and add MarshalJSON for CloudConfigFile

- Removed base64 encoding handling from UnmarshalJSON for clarity.
- Added MarshalJSON implementation to ensure proper JSON serialization of CloudConfigFile.

alexlovelltroy committed:
feat(bss): implement BootParams and boot script generation with HTTP handlers

- Added BootParams struct to define boot parameters including kernel, initrd, root filesystem, and cloud-init server configurations.
- Implemented GenerateBootScript function to create iPXE boot scripts based on BootParams.
- Created BootParamsHandler for handling HTTP requests to create, update, retrieve, and generate boot scripts.
- Introduced MemoryStore for in-memory storage of boot parameters with versioning.
- Added comprehensive unit tests for BootParams, boot script generation, and HTTP handlers to ensure functionality and reliability.

alexlovelltroy committed:
feat(bss): Add the capability for cloud-init to act as BSS

- Updated go.mod to include the new dependency for UUID generation.
- Refactored import paths to use the new smdclient package structure.
- Introduced BootParamsV1 struct for backward compatibility with the old API specification.
- Enhanced boot parameters validation and merging logic.
- Implemented new HTTP handlers for creating and managing boot parameters.
- Added comprehensive unit tests for the new features and refactored existing tests for better coverage.

alexlovelltroy committed:
feat(quackstore): implement unique instance ID generation

- Added functionality to generate a unique instance ID in the format "i-XXXXXX" using a random 6-digit hexadecimal string.
- Updated the comment to clarify the instance ID generation process.

alexlovelltroy committed:
Merging main into bss branch

alexlovelltroy committed:
feat(bss): add BSS API endpoints for boot parameters and scripts

- Introduced new BSS API endpoints for managing boot parameters and generating boot scripts.
- Added handlers for creating, updating, and retrieving boot parameters.
- Implemented routing for BSS endpoints in the main application.
- Updated go.mod to include new dependencies for BSS functionality.
- Enhanced documentation for the new API endpoints with Swagger annotations.

synackd commented:
I think that this is meant to be a solution to BSS's current lack of group name support when using the Postgres backend, but I'm a bit worried that by adding BSS functionality to cloud-init, we are stepping towards distributed monoliths instead of microservices, which I believe is the same reason we removed discovery from SMD.

Is there an important use case for having cloud-init also provide boot parameter management and script generation? Just trying to understand why we're duplicating BSS functionality here versus refactoring BSS.

davidallendj commented:
I agree with @synackd here. I believed we had this discussion before and IIRC I think before we were certainly against duplicating the same functionality in multiple microservices. Is the intent here to eventually phase out BSS entirely? Are we not packing too many features in cloud-init alone? I'm wondering if adding this will move cloud-init in the same direction as SMD.

alexlovelltroy commented:
I have confined the BSS portions to a `pkg/` so we can decide later if we want to maintain a separate BSS or if we want to integrate cloud-init and bss.

rustydb commented:
qq, are the functions of BSS that OpenCHAMI was leveraging confined to just bootparameters and bootscripts?

alexlovelltroy commented:
@rustydb 

There are four endpoints described in the [swaggerfile](https://github.com/OpenCHAMI/bss/blob/main/api/swagger.yaml).  At this point, OpenCHAMI is only leveraging the two concerned with bootparameters and bootscripts.  It didn't look to us like there was added value in the dumpstate or hosts endpoints.

1.    `/boot/v1/bootscript` 
2.    `/boot/v1/bootparameters`
3.    `/boot/v1/hosts`
4.    `/boot/v1/dumpstate`

I'm not closed to the idea that we might leverage more of the API at some point.  We simply haven't needed them.  Are we missing something?

synackd commented:
I think reimplementing an API-compliant version of BSS that allows the functionality that we want to implement is probably the way to go, but I'm wondering if cloud-init is the right place for it. 

Are we sure we want to combine boot parameter handling and post-boot configuration into a single scope?

rustydb commented:
> @rustydb
> 
> There are four endpoints described in the [swaggerfile](https://github.com/OpenCHAMI/bss/blob/main/api/swagger.yaml). At this point, OpenCHAMI is only leveraging the two concerned with bootparameters and bootscripts. It didn't look to us like there was added value in the dumpstate or hosts endpoints.

Makes sense. FWIW I'm looking for any gaps, anything that BSS does today that isn't covered here and whether that matters.

That really just leaves the pre-signed URLs (for private S3 buckets) and spire-tokens, which I don't see mentioned in the PR. I assume those aren't important to OpenCHAMI either.

alexlovelltroy commented:
We're not using pre-signed urls or spire tokens at this point in OpenCHAMI.  I'm not convinced that either add security given the unencrypted/unauthenticated nature of iPXE.  However, if we choose to go down this path, we may need to open both assertions up for discussion.

Pre-signed urls aren't necessary for LANL at this time because none of our system images, kernels, or initrd contain any secrets and are on segregated networks.  If OpenCHAMI needs to support authenticated access to S3 resources, we'll need to update this project.

I'll defer to @yogi4 who has worked a lot more with spire, but I think we'd prefer not to embed the spire join token in the kernel commandline and instead leverage the tpm attestation process to initialize spire.

alexlovelltroy committed:
Merge branch 'main' into alovelltroy/bss-in-cloud-init

alexlovelltroy commented:
Looks like there is a boot/v1/endpoint-history that will provide information about the last time each node retrieved their bootscript.  That sounds pretty useful

Masber commented:
Just sharing my personal opinion about cloud-init and BSS:

These services serve different purposes and operate at different stages. BSS handles boot-time configuration (setting kernel parameters or delivering initrd, etc) essentially shaping how the system starts. cloud-init, by contrast, is about runtime configuration once the OS is up (e.g., user setup, networking, mount filesystems, package installs).

I agree with @synackd’s point, but I’d also emphasize that cloud-init isn't a given (not every OCHAMI user will use it). Runtime config is complex and often handled differently across environments (cloud-init, Ansible, custom scripts, etc.).

alexlovelltroy commented:
I agree that we don't want to force people into using cloud-init if they don't want to.

I'm updating this branch/PR to isolate the BSS API in a separate executable with a separate build.  We can review it once that's functional

alexlovelltroy committed:
feat: Add netboot server implementation and API endpoints for boot parameters

- Introduced a new netboot server in cmd/netboot-server/main.go.
- Updated .goreleaser.yaml to include netboot-server build configuration.
- Added API endpoints for creating, retrieving, and updating boot parameters in docs/swagger.yaml and docs/docs.go.
- Defined BootParams schema in swagger documentation and Go code.

rainest reviewed (commented):
This could basically just stuff BSS under a router subsection, correct? As-is it has multiple BSS endpoints under the root, but I think if we group those and adhere to code organization constraints, keeping these in a single repo/binary is fine.

Most everything this adds resides under `pkg/bss`. The only parts that need to interact with both are the main/cmd and the router, and we can design those such that they keep the components separate, and that we could, if desired, split them into actual separate repos later.

We'd need to be careful about how we code shared components, since it wouldn't be _enforced_ by split repos, but it's doable. There are some bits in the PR as written that don't keep that separation clean (see the in-context comments).

I think you're also blocked from just ignoring the cloud-init bit--it looks like the cluster defaults initialization creates de-facto required configuration specific to cloud-init, and ideally we wouldn't have that.

Arguments for are:
- These are similar enough in use and needs: they're both basically infrequent human write, frequent machine read config stores that communicate with SMD a bit, and they both need to be accessible to machines before those machines are fully configured.
- Hey, it's an excuse to get away from the BSS codebase! If a simpler version of that is sufficient, may as well.

If we do want to separate it, we should have a separate repo rather than a separate binary: pinning them to the same module versions and needing multiple Containerfiles is annoying. SMDClient is probably a reasonable module-only repo, and not re-implementing that across services makes sense anyway.

alexlovelltroy commented on cmd/cloud-init-server/main.go:
```diff
@@ -236,3 +266,27 @@ func initCiAdminRouter(router chi.Router, handler *CiHandler) {
 	})
 }
+
+func initBssRouter(router chi.Router, handler *CiHandler, bssStorage bss.Store) {
```
I completely agree.  Based on other feedback, I'm working on a separate binary to serve the BSS portions.  The new signature is at: https://github.com/OpenCHAMI/cloud-init/blob/3dbbefb489f6c9612af7edf9143b3eef5fdcc896/cmd/netboot-server/main.go#L129

Cleaning up that struct is on my mental todo list for cloud-init as well.

alexlovelltroy committed:
feat(bss): add CmdLine field to BootParams for kernel command line configuration

alexlovelltroy committed:
feat: enhance netboot server with versioning and QuackStore integration

- Added QuackStore for persistent storage of boot parameters, supporting versioning and group templates.
- Implemented API endpoints for managing boot parameters, including creation, retrieval, updating, and versioning.
- Introduced a version command to display build information and server version.
- Updated .goreleaser.yaml to support new archive formats (zip and tar.gz).
- Enhanced documentation with new fields and API specifications in swagger and Go code.
- Added tests for QuackStore functionality to ensure data integrity and operations.

alexlovelltroy committed:
fix: update archive formats in GoReleaser configuration and modify snapshot build command

rainest reviewed (changes requested):
Comments on https://github.com/OpenCHAMI/cloud-init/pull/71/commits/8c4c2b8c5956d3c947e674c664f092a74eafa8e9 mostly re some organization changes to better allow alternate data stores and avoiding bash for integration tests.

alexlovelltroy commented:
waiting to rebase on top of the quackstore PR (#68) once that merges.
